### DISCUSSION TITLE
Tracing is seperate in Vnodes.

### DISCUSSION THREAD

**Discussion by Terni** [apache/skywalking#9262] [Q&A] [answered]
Hi skywalking team.
I have a problem with the tracing sequence. VNode is separate, I don't know how to fix it.
![image](https://user-images.githubusercontent.com/1767477/175014491-bccc5f3a-d948-467e-b8dc-958608a30c2e.png)

![image](https://user-images.githubusercontent.com/1767477/175014620-3a699520-5d99-43d7-a87c-e2e877b0c827.png)

Sender with ExitSegmentContext:

        public virtual void BeforeTransportSend(object value)
        {
            if (value is not RabbitMqSendContext mqSendContext)
            {
                return;
            }

            try
            {
                string operationName = value.GetType().IsGenericType
                    ? value.GetType().GetGenericArguments().First().Name
                    : mqSendContext.Exchange;

                string networkAddrerss = mqSendContext.SourceAddress?.ToString() ?? string.Empty;
                var brokerAddress = new BrokerAddress(MqComponent, networkAddrerss);

                var headerCollection = new TextCarrierHeaderCollection(new Dictionary<string, string>());
                SegmentContext exitSegmentContext = _tracingContext.CreateExitSegmentContext($"Send_{operationName}", networkAddrerss, headerCollection);

                var correlationId = mqSendContext.CorrelationId?.ToString() ?? string.Empty;
                
                exitSegmentContext.Span.Component = GetComponent(brokerAddress, true);
                exitSegmentContext.Span.SpanLayer = SpanLayer.MQ;
                exitSegmentContext.Span.Peer = brokerAddress.Endpoint;
                exitSegmentContext.Span.AddTag(Tags.MQ_TOPIC, operationName);
                exitSegmentContext.Span.AddTag(Tags.MQ_BROKER, HostBroker);
                exitSegmentContext.Span.AddTag("CorrelationId", correlationId);
                exitSegmentContext.Span.AddTag("MessageId", mqSendContext.MessageId?.ToString() ?? string.Empty);
                exitSegmentContext.Span.AddTag("RequestId", mqSendContext.RequestId?.ToString() ?? string.Empty);
                
                if (mqSendContext.MessageId != null)
                {
                    _contexts[mqSendContext.MessageId.Value] = exitSegmentContext;
                }

                if (!headerCollection.AnySafe())
                {
                    return;
                }

                foreach (var (key, headerVal) in headerCollection)
                {
                    mqSendContext.Headers.Set(key, headerVal);
                }
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Failed to process MassTransit diagnostic event!");
            }
        }

        public virtual void AfterTransportSend(object value)
        {
            if (value is not RabbitMqSendContext mqSendContext)
            {
                return;
            }
            
            if (mqSendContext.MessageId != null && _contexts.TryRemove(mqSendContext.MessageId.Value, out var exitContext))
            {
                _tracingContext.Release(exitContext);
            }
        }

Receiver with EntrySegmentContext:

        public virtual void BeforeConsumerHandle(object value)
        {
            if (value is not ConsumeContext consumeContext)
            {
                return;
            }

            try
            {
                //only string headers are supported by SW
                IEnumerable<KeyValuePair<string, string>> headerPair = consumeContext.Headers
                    .GetAll()
                    .Select(header => new KeyValuePair<string, string>(header.Key, header.Value.ToString()));
                var headerCollection = new TextCarrierHeaderCollection(headerPair);
                string networkAddress = consumeContext.SourceAddress?.ToString() ?? string.Empty;
                var brokerAddress = new BrokerAddress(MqComponent,networkAddress);
                
                string operationName = value.GetType().IsGenericType ? value.GetType().GetGenericArguments().First().Name : networkAddress;
                SegmentContext entrySegmentContext = _tracingContext.CreateEntrySegmentContext($"Handle_{operationName}", headerCollection);
                
                entrySegmentContext.Span.Component = GetComponent(brokerAddress, false);
                entrySegmentContext.Span.SpanLayer = SpanLayer.MQ;
                entrySegmentContext.Span.Peer = brokerAddress.Endpoint;
                entrySegmentContext.Span.AddTag("CorrelationId", consumeContext.CorrelationId?.ToString() ?? string.Empty);
                entrySegmentContext.Span.AddTag("MessageId", consumeContext.MessageId?.ToString() ?? string.Empty);
                entrySegmentContext.Span.AddTag("RequestId", consumeContext.RequestId?.ToString() ?? string.Empty);
                
                if (consumeContext.MessageId != null)
                {
                    _contexts[consumeContext.MessageId.Value] = entrySegmentContext;
                }
                
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Failed to process MassTransit diagnostic event!");
            }
        }

And last version skywalking 9.1.0 in server-oap. I saw this error 

```
2022-06-22 10:05:27,149 org.apache.skywalking.oap.server.receiver.trace.provider.handler.v8.grpc.TraceSegmentReportServiceHandler 75 [grpcServerPool-1-thread-15] ERROR [] - [9.1.0-f1f519c] null
java.lang.NullPointerException: null
at org.apache.skywalking.oap.server.core.source.EndpointRelation.prepare(EndpointRelation.java:101) ~[server-core-9.1.0.jar:9.1.0]
at org.apache.skywalking.oap.server.core.analysis.DispatcherManager.forward(DispatcherManager.java:59) ~[server-core-9.1.0.jar:9.1.0]
at org.apache.skywalking.oap.server.core.source.SourceReceiverImpl.receive(SourceReceiverImpl.java:36) ~[server-core-9.1.0.jar:9.1.0]
at org.apache.skywalking.oap.server.analyzer.provider.trace.parser.listener.EndpointDepFromCrossThreadAnalysisListener.lambda$build$0(EndpointDepFromCrossThreadAnalysisListener.java:130) ~[agent-analyzer-9.1.0.jar:9.1.0]
at java.util.ArrayList.forEach(Unknown Source) ~[?:?]
at org.apache.skywalking.oap.server.analyzer.provider.trace.parser.listener.EndpointDepFromCrossThreadAnalysisListener.build(EndpointDepFromCrossThreadAnalysisListener.java:124) ~[agent-analyzer-9.1.0.jar:9.1.0
```

Do I have to set something somewhere?

**Comment by wu-sheng** [comment, marked as answer]
This seems your private codes on your side. You have the source codes we don't.

About `VNode`, there is a very clear explanation in the doc about what it is.

**Reply by Terni** [reply]
Ok I understood. 
https://github.com/apache/skywalking/blob/master/docs/en/FAQ/vnode.md
I think about it.

**Reply by xlq20080808** [reply]
I have the same problem, have you found the cause of the NPE?

**Reply by Terni** [reply]
I solved it. The problem was pairing to a different type. I had to remove (Component, SpanLayer, Peer). There is some documentation, but it is very weak. I hope there will be documentation for CreateExitSegmentContext in the future.

**Reply by xlq20080808** [reply]
I solved this problem by used the main branch code.This is a bug about MQ
https://github.com/apache/skywalking/issues/9297#issuecomment-1173057480 and fixed by https://github.com/apache/skywalking/pull/9300

**Reply by Terni** [reply]
Thank you. I knew it but nobody help me.
